File: tests/src/test/java/com/github/klee0kai/stone/test/cache/EarthSwitchCacheTests.java
package com.github.klee0kai.stone.test.cache;

import com.github.klee0kai.stone.Stone;
import com.github.klee0kai.test.di.swcache.SwitchCacheComponent;
import com.github.klee0kai.test.mowgli.earth.Mountain;
import com.github.klee0kai.test.mowgli.earth.River;
import org.junit.jupiter.api.Test;

import java.lang.ref.WeakReference;

import static java.lang.Thread.sleep;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNull;

public class EarthSwitchCacheTests {


    @Test
    public void allToWeakTest() {
        //Given
        SwitchCacheComponent di = Stone.createComponent(SwitchCacheComponent.class);
        WeakReference<Mountain> mountain = new WeakReference<>(di.earth().mountainStrong());

        //When
        di.allWeak();
        System.gc();

        //Then
        assertNull(mountain.get());
    }

    @Test
    public void strongToWeakTest() {
        //Given
        SwitchCacheComponent di = Stone.createComponent(SwitchCacheComponent.class);
        WeakReference<Mountain> mountainStrong = new WeakReference<>(di.earth().mountainStrong());
        WeakReference<Mountain> mountainSoft = new WeakReference<>(di.earth().mountainSoft());

        //When
        di.strongToWeak();
        System.gc();

        //Then
        assertNull(mountainStrong.get());
        assertNotNull(mountainSoft.get());
    }

    @Test
    public void weakToStrongFewMillisTest() throws InterruptedException {
        //Given
        SwitchCacheComponent di = Stone.createComponent(SwitchCacheComponent.class);
        WeakReference<Mountain> mountainWeak = new WeakReference<>(di.earth().mountainWeak());

        //When
        di.allStrongFewMillis();
        System.gc();

        //Then: can't GC
        assertNotNull(mountainWeak.get());

        //When: after few millis
        sleep(110);
        System.gc();

        //Then: can GC
        assertNull(mountainWeak.get());
    }


    @Test
    public void mountainToWeakTest() {
        //Given
        SwitchCacheComponent di = Stone.createComponent(SwitchCacheComponent.class);
        WeakReference<Mountain> mountain = new WeakReference<>(di.earth().mountainStrong());
        WeakReference<River> river = new WeakReference<>(di.earth().riverSoft());

        //When
        di.mountainToWeak();
        System.gc();

        //Then
        assertNull(mountain.get());
        assertNotNull(river.get());
    }


}

File: tests/src/test/java/com/github/klee0kai/stone/test/cache/EarthSwitchCacheTests.kt
package com.github.klee0kai.stone.test.cache

import com.github.klee0kai.test.di.swcache.SwitchCacheComponentStoneComponent
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertNotNull
import org.junit.jupiter.api.assertNull
import java.lang.ref.WeakReference

class EarthSwitchCacheTests {

    @Test
    fun allToWeakTest() {
        //Given
        val DI = SwitchCacheComponentStoneComponent()
        val mountain = WeakReference(DI.earth().mountainStrong())

        //When
        DI.allWeak()
        System.gc()

        //Then
        assertNull(mountain.get())
    }

    @Test
    fun strongToWeakTest() {
        //Given
        val DI = SwitchCacheComponentStoneComponent()
        val mountainStrong = WeakReference(DI.earth().mountainStrong())
        val mountainSoft = WeakReference(DI.earth().mountainSoft())

        //When
        DI.strongToWeak()
        System.gc()

        //Then
        assertNull(mountainStrong.get())
        assertNotNull(mountainSoft.get())
    }

    @Test
    @Throws(InterruptedException::class)
    fun weakToStrongFewMillisTest() {
        //Given
        val DI = SwitchCacheComponentStoneComponent()
        val mountainWeak = WeakReference(DI.earth().mountainWeak())

        //When
        DI.allStrongFewMillis()
        System.gc()

        //Then: can't GC
        assertNotNull(mountainWeak.get())

        //When: after few millis
        Thread.sleep(110)
        System.gc()

        //Then: can GC
        assertNull(mountainWeak.get())
    }


    @Test
    fun mountainToWeakTest() {
        //Given
        val DI = SwitchCacheComponentStoneComponent()
        val mountain = WeakReference(DI.earth().mountainStrong())
        val river = WeakReference(DI.earth().riverSoft())

        //When
        DI.mountainToWeak()
        System.gc()

        //Then
        assertNull(mountain.get())
        assertNotNull(river.get())
    }

}
